Components/SoarDebug/DebugTask.cpp

```diff
@@ -39,15 +68,97 @@ void DebugTask::InitTask()
 	SOAR_ASSERT(rtValue == pdPASS, "UARTTask::InitTask() - xTaskCreate() failed");
 }
 
-// TODO: Only run thread when appropriate GPIO pin pulled HIGH
+// TODO: Only run thread when appropriate GPIO pin pulled HIGH (or by define)
+/**
+ *	@brief Runcode for the DebugTask
+ */
 void DebugTask::Run(void * pvParams)
 {
-    uint32_t prevWakeTime = osKernelSysTick();
-    //uint8_t buffer = 0x00;
+	// Arm the interrupt
+	ReceiveData();
 
 	while (1) {
-		osDelayUntil(&prevWakeTime, DEBUG_TASK_PERIOD);
+		Command cm;
+
+		//Wait forever for a command
+		qEvtQueue->ReceiveWait(cm);
+
+		//Process the command
+		if(cm.GetCommand() == DATA_COMMAND && cm.GetTaskCommand() == EVENT_DEBUG_RX_COMPLETE) {
+			HandleDebugMessage((const char*)debugBuffer);
+		}
+
+		cm.Reset();
+	}
+}
+
+/**
+ * @brief Handles debug messages, assumes msg is null terminated
+ * @param msg Message to read, must be null termianted
+ */
+void DebugTask::HandleDebugMessage(const char* msg)
+{
+	if (strcmp(msg, "sysreset") == 0) {
+		// Reset the system
+		SOAR_ASSERT(false, "System reset requested");
+	}
+	else if (strcmp(msg, "sysinfo") == 0) {
+		// Print message
+		SOAR_PRINT("\n-- SOAR System Info --\n");
+		SOAR_PRINT("Current System Heap Use: %d Bytes\n", xPortGetFreeHeapSize());
+		SOAR_PRINT("Lowest Ever Heap Size: %d Bytes\n\n", xPortGetMinimumEverFreeHeapSize());
+	}
+	else if (strcmp(msg, "blinkled") == 0) {
+		// Print message
+		SOAR_PRINT("Debug 'LED blink' command requested\n");
+		GPIO::LED1::On();
+		// TODO: Send to HID task to blink LED, this shouldn't delay
+	}
+	else {
+		// Single character command
+		switch (msg[0]) {
+		default:
+			SOAR_PRINT("Debug, unknown command: %s\n", msg);
+			break;
+		}
+	}
+
+	//We've read the data, clear the buffer
+	debugMsgIdx = 0;
+	isDebugMsgReady = false;
+}
 
-		//HAL_UART_Receive(&huart5, &buffer, 1, 1000); // This should be in UART Task (UART Task's job is to poll these buffers)
+/**
+ * @brief Receive data, currently receives by arming interrupt
+ */
+bool DebugTask::ReceiveData()
+{
+	HAL_UART_Receive_IT(SystemHandles::UART_Debug, &debugRxChar, 1);
+	return true;
+}
+
+/**
+ * @brief Receive data to the buffer
+ * @return Whether the debugBuffer is ready or not
+ */
+void DebugTask::InterruptRxData()
+{
+	// If we already have an unprocessed debug message, ignore this byte
+	if (!isDebugMsgReady) {
+		// Check byte for end of message - note if using termite you must turn on append CR
+		if (debugRxChar == '\r' || debugRxChar == '\0' || debugMsgIdx == DEBUG_RX_BUFFER_SZ_BYTES) {
+			// Null terminate and process
+			debugBuffer[debugMsgIdx++] = '\0';
+
+			// Notify the debug task
+			Command cm(DATA_COMMAND, EVENT_DEBUG_RX_COMPLETE);
+			qEvtQueue->SendFromISR(cm);
```

Comment:
If this fails, we need to clear the buffer and re-arm the interrupt (otherwise we'll lose data, even though this should really never happen)